File: src/smartcar_safety/smartcar_safety/guard.py
"""ROS-independent fail-closed velocity safety decisions."""
import math


def _finite_float(name, value):
    value = float(value)
    if not math.isfinite(value):
        raise ValueError(f"{name} must be finite")
    return value


def require_positive_finite(name, value):
    value = _finite_float(name, value)
    if value <= 0.0:
        raise ValueError(f"{name} must be greater than zero")
    return value


def require_nonnegative_finite(name, value):
    value = _finite_float(name, value)
    if value < 0.0:
        raise ValueError(f"{name} must be nonnegative")
    return value


def validate_publish_frequency(value):
    return require_positive_finite("publish_frequency_hz", value)


class SafetyGuard:
    """Track received inputs and decide whether a velocity command is safe."""

    def __init__(
        self,
        command_timeout_sec=0.30,
        scan_timeout_sec=0.35,
        odom_timeout_sec=0.35,
        minimum_voltage=0.0,
        voltage_timeout_sec=1.0,
        max_linear_speed_mps=0.30,
        require_scan=True,
        require_odom=True,
        raw_odom_timeout_sec=0.25,
        require_raw_odom=True,
        depth_points_timeout_sec=0.50,
        require_depth_points=False,
    ):
        self.command_timeout_sec = require_positive_finite(
            "command_timeout_sec", command_timeout_sec)
        self.scan_timeout_sec = require_positive_finite(
            "scan_timeout_sec", scan_timeout_sec)
        self.odom_timeout_sec = require_positive_finite(
            "odom_timeout_sec", odom_timeout_sec)
        self.raw_odom_timeout_sec = require_positive_finite(
            "raw_odom_timeout_sec", raw_odom_timeout_sec)
        self.depth_points_timeout_sec = require_positive_finite(
            "depth_points_timeout_sec", depth_points_timeout_sec)
        self.minimum_voltage = require_nonnegative_finite(
            "minimum_voltage", minimum_voltage)
        self.voltage_timeout_sec = require_positive_finite(
            "voltage_timeout_sec", voltage_timeout_sec)
        self.max_linear_speed_mps = require_positive_finite(
            "max_linear_speed_mps", max_linear_speed_mps)
        self.require_scan = bool(require_scan)
        self.require_odom = bool(require_odom)
        self.require_raw_odom = bool(require_raw_odom)
        self.require_depth_points = bool(require_depth_points)

        self.command_received_at = None
        self.command_invalid = False
        self.command_speed_limit_exceeded = False
        self.scan_received_at = None
        self.odom_received_at = None
        self.raw_odom_received_at = None
        self.depth_points_received_at = None
        self.voltage_received_at = None
        self.voltage = None
        self.emergency_stop = False

    def mark_command(self, receipt_time_sec, linear_x):
        if self.command_speed_limit_exceeded:
            return False
        try:
            linear_x = float(linear_x)
        except (TypeError, ValueError, OverflowError):
            self.mark_command_invalid()
            return False
        if not math.isfinite(linear_x):
            self.mark_command_invalid()
            return False
        if abs(linear_x) > self.max_linear_speed_mps:
            self.command_invalid = False
            self.command_speed_limit_exceeded = True
            return False
        self.command_received_at = float(receipt_time_sec)
        self.command_invalid = False
        self.command_speed_limit_exceeded = False
        return True

    def mark_command_invalid(self):
        self.command_invalid = True

    def mark_scan(self, receipt_time_sec):
        self.scan_received_at = float(receipt_time_sec)

    def mark_odom(self, receipt_time_sec):
        self.odom_received_at = float(receipt_time_sec)

    def mark_raw_odom(self, receipt_time_sec):
        self.raw_odom_received_at = float(receipt_time_sec)

    def mark_depth_points(self, receipt_time_sec):
        self.depth_points_received_at = float(receipt_time_sec)

    def mark_voltage(self, voltage, receipt_time_sec):
        self.voltage = float(voltage)
        self.voltage_received_at = float(receipt_time_sec)

    def set_emergency_stop(self, enabled):
        """Latch a stop until a caller explicitly clears it with False."""
        self.emergency_stop = bool(enabled)

    def clear_command_speed_limit_fault(self):
        self.command_speed_limit_exceeded = False

    @staticmethod
    def _fresh(receipt_time_sec, now_sec, timeout_sec):
        if receipt_time_sec is None:
            return False
        age_sec = float(now_sec) - receipt_time_sec
        return 0.0 <= age_sec <= timeout_sec

    @staticmethod
    def _result(allowed, reason):
        return {"allowed": allowed, "reason": reason}

    def command_is_fresh(self, now_sec):
        return self._fresh(
            self.command_received_at, now_sec, self.command_timeout_sec)

    def evaluate(self, now_sec):
        """Return a dict containing the fail-closed decision and its reason."""
        if self.emergency_stop:
            return self._result(False, "emergency_stop")

        if self.command_speed_limit_exceeded:
            return self._result(False, "command_speed_limit_exceeded")

        if self.command_invalid:
            return self._result(False, "command_invalid")

        if self.command_received_at is None:
            return self._result(False, "command_missing")
        if not self._fresh(self.command_received_at, now_sec, self.command_timeout_sec):
            return self._result(False, "command_stale")

        if self.require_scan:
            if self.scan_received_at is None:
                return self._result(False, "scan_missing")
            if not self._fresh(self.scan_received_at, now_sec, self.scan_timeout_sec):
                return self._result(False, "scan_stale")

        if self.require_odom:
            if self.odom_received_at is None:
                return self._result(False, "odom_missing")
            if not self._fresh(self.odom_received_at, now_sec, self.odom_timeout_sec):
                return self._result(False, "odom_stale")

        if self.require_raw_odom:
            if self.raw_odom_received_at is None:
                return self._result(False, "raw_odom_missing")
            if not self._fresh(
                self.raw_odom_received_at, now_sec, self.raw_odom_timeout_sec
            ):
                return self._result(False, "raw_odom_stale")

        if self.require_depth_points:
            if self.depth_points_received_at is None:
                return self._result(False, "depth_points_missing")
            if not self._fresh(
                self.depth_points_received_at,
                now_sec,
                self.depth_points_timeout_sec,
            ):
                return self._result(False, "depth_points_stale")

        if self.minimum_voltage > 0.0:
            if self.voltage is None:
                return self._result(False, "voltage_missing")
            if not self._fresh(
                self.voltage_received_at, now_sec, self.voltage_timeout_sec
            ):
                return self._result(False, "voltage_stale")
            if not math.isfinite(self.voltage):
                return self._result(False, "voltage_invalid")
            if self.voltage < self.minimum_voltage:
                return self._result(False, "voltage_low")

        return self._result(True, "ok")

File: src/smartcar_safety/src/guard.cpp
#include "smartcar_safety/guard.hpp"

#include <cmath>
#include <stdexcept>
#include <string>

namespace smartcar_safety {

SafetyGuard::SafetyGuard(double cmd_timeout, double scan_timeout,
                         double odom_timeout, double raw_odom_timeout,
                         double min_voltage, bool require_scan,
                         bool require_odom, bool require_raw_odom)
    : cmd_timeout_(cmd_timeout),
      scan_timeout_(scan_timeout),
      odom_timeout_(odom_timeout),
      raw_odom_timeout_(raw_odom_timeout),
      min_voltage_(min_voltage),
      require_scan_(require_scan),
      require_odom_(require_odom),
      require_raw_odom_(require_raw_odom),
      cmd_at_(NAN),
      scan_at_(NAN),
      odom_at_(NAN),
      raw_odom_at_(NAN),
      voltage_at_(NAN),
      voltage_(NAN),
      cmd_invalid_(false),
      estop_(false),
      cmd_at_set_(false),
      scan_at_set_(false),
      odom_at_set_(false),
      raw_odom_at_set_(false),
      voltage_at_set_(false) {
  if (!std::isfinite(cmd_timeout) || cmd_timeout <= 0.0) {
    throw std::invalid_argument("command_timeout_sec must be positive finite");
  }
  if (!std::isfinite(scan_timeout) || scan_timeout <= 0.0) {
    throw std::invalid_argument("scan_timeout_sec must be positive finite");
  }
  if (!std::isfinite(odom_timeout) || odom_timeout <= 0.0) {
    throw std::invalid_argument("odom_timeout_sec must be positive finite");
  }
  if (!std::isfinite(raw_odom_timeout) || raw_odom_timeout <= 0.0) {
    throw std::invalid_argument(
        "raw_odom_timeout_sec must be positive finite");
  }
  if (!std::isfinite(min_voltage) || min_voltage < 0.0) {
    throw std::invalid_argument("minimum_voltage must be nonnegative finite");
  }
}

void SafetyGuard::mark_command(double now_sec) {
  cmd_at_ = now_sec;
  cmd_at_set_ = true;
  cmd_invalid_ = false;
}

void SafetyGuard::mark_command_invalid() { cmd_invalid_ = true; }

void SafetyGuard::mark_scan(double now_sec) {
  scan_at_ = now_sec;
  scan_at_set_ = true;
}

void SafetyGuard::mark_odom(double now_sec) {
  odom_at_ = now_sec;
  odom_at_set_ = true;
}

void SafetyGuard::mark_raw_odom(double now_sec) {
  raw_odom_at_ = now_sec;
  raw_odom_at_set_ = true;
}

void SafetyGuard::mark_voltage(float voltage, double now_sec) {
  voltage_ = voltage;
  voltage_at_ = now_sec;
  voltage_at_set_ = true;
}

void SafetyGuard::set_emergency_stop(bool enabled) { estop_ = enabled; }

bool SafetyGuard::command_is_fresh(double now_sec) const {
  return fresh(cmd_at_, now_sec, cmd_timeout_);
}

SafetyVerdict SafetyGuard::evaluate(double now_sec) const {
  if (estop_) {
    return result(false, "emergency_stop");
  }

  if (cmd_invalid_) {
    return result(false, "command_invalid");
  }

  if (!cmd_at_set_) {
    return result(false, "command_missing");
  }
  if (!fresh(cmd_at_, now_sec, cmd_timeout_)) {
    return result(false, "command_stale");
  }

  if (require_scan_) {
    if (!scan_at_set_) {
      return result(false, "scan_missing");
    }
    if (!fresh(scan_at_, now_sec, scan_timeout_)) {
      return result(false, "scan_stale");
    }
  }

  if (require_odom_) {
    if (!odom_at_set_) {
      return result(false, "odom_missing");
    }
    if (!fresh(odom_at_, now_sec, odom_timeout_)) {
      return result(false, "odom_stale");
    }
  }

  if (require_raw_odom_) {
    if (!raw_odom_at_set_) {
      return result(false, "raw_odom_missing");
    }
    if (!fresh(raw_odom_at_, now_sec, raw_odom_timeout_)) {
      return result(false, "raw_odom_stale");
    }
  }

  if (min_voltage_ > 0.0) {
    if (!voltage_at_set_) {
      return result(false, "voltage_missing");
    }
    if (!std::isfinite(voltage_)) {
      return result(false, "voltage_invalid");
    }
    if (voltage_ < min_voltage_) {
      return result(false, "voltage_low");
    }
  }

  return result(true, "ok");
}

bool SafetyGuard::fresh(double arrival, double now, double timeout) {
  if (std::isnan(arrival)) {
    return false;
  }
  double age = now - arrival;
  return age >= 0.0 && age <= timeout;
}

SafetyVerdict SafetyGuard::result(bool allowed, const std::string &reason) {
  return SafetyVerdict{allowed, reason};
}

}  // namespace smartcar_safety
